cmd/notification.go

```diff
@@ -0,0 +1,634 @@
+			delete(sys.bucketRemoteTargetRulesMap, bucketName)
+		}
+	}
+}
+
+func (sys *NotificationSys) send(bucketName string, eventData event.Event, targetIDs ...event.TargetID) map[event.TargetID]error {
+	errMap := sys.targetList.Send(eventData, targetIDs...)
+	for targetID := range errMap {
+		if sys.RemoteTargetExist(bucketName, targetID) {
+			sys.RemoveRemoteTarget(bucketName, targetID)
+		}
+	}
+
+	return errMap
+}
+
+// Send - sends event data to all matching targets.
+func (sys *NotificationSys) Send(args eventArgs) map[event.TargetID]error {
+	sys.RLock()
+	targetIDSet := sys.bucketRulesMap[args.BucketName].Match(args.EventName, args.Object.Name)
+	sys.RUnlock()
+	if len(targetIDSet) == 0 {
+		return nil
+	}
+
+	targetIDs := targetIDSet.ToSlice()
+	return sys.send(args.BucketName, args.ToEvent(), targetIDs...)
+}
+
+// NewNotificationSys - creates new notification system object.
+func NewNotificationSys(config *serverConfig, endpoints EndpointList) (*NotificationSys, error) {
+	targetList, err := getNotificationTargets(config)
+	if err != nil {
+		return nil, err
+	}
+
+	peerRPCClientMap := makeRemoteRPCClients(endpoints)
+
+	// bucketRulesMap/bucketRemoteTargetRulesMap are initialized by NotificationSys.Init()
+	return &NotificationSys{
+		targetList:                 targetList,
+		bucketRulesMap:             make(map[string]event.RulesMap),
+		bucketRemoteTargetRulesMap: make(map[string]map[event.TargetID]event.RulesMap),
+		peerRPCClientMap:           peerRPCClientMap,
+	}, nil
+}
+
+type eventArgs struct {
+	EventName  event.Name
+	BucketName string
+	Object     ObjectInfo
+	ReqParams  map[string]string
+	Host       string
+	Port       string
+	UserAgent  string
+}
+
+// ToEvent - converts to notification event.
+func (args eventArgs) ToEvent() event.Event {
+	getOriginEndpoint := func() string {
+		host := globalMinioHost
+		if host == "" {
+			// FIXME: Send FQDN or hostname of this machine than sending IP address.
+			host = localIP4.ToSlice()[0]
+		}
+
+		return fmt.Sprintf("%s://%s:%s", getURLScheme(globalIsSSL), host, globalMinioPort)
+	}
+
+	creds := globalServerConfig.GetCredential()
+	eventTime := UTCNow()
+	uniqueID := fmt.Sprintf("%X", eventTime.UnixNano())
+
+	newEvent := event.Event{
+		EventVersion:      "2.0",
+		EventSource:       "minio:s3",
+		AwsRegion:         globalServerConfig.GetRegion(),
+		EventTime:         eventTime.Format(event.AMZTimeFormat),
+		EventName:         args.EventName,
+		UserIdentity:      event.Identity{creds.AccessKey},
+		RequestParameters: args.ReqParams,
+		ResponseElements: map[string]string{
+			"x-amz-request-id":        uniqueID,
+			"x-minio-origin-endpoint": getOriginEndpoint(), // Minio specific custom elements.
+		},
+		S3: event.Metadata{
+			SchemaVersion:   "1.0",
+			ConfigurationID: "Config",
+			Bucket: event.Bucket{
+				Name:          args.BucketName,
+				OwnerIdentity: event.Identity{creds.AccessKey},
+				ARN:           "arn:aws:s3:::" + args.BucketName,
```

Comment:
This can be `bucketARNPrefix + args.BucketName` ?

Reply:
Done.